File: jj/faultdiag-source-20260906/annotation-app/frontend/src/types/annotation.ts
export type AnnotationLabel = 'normal' | 'abnormal'
export type AnnotationExportScope = 'current_filter' | 'all'
export type AnnotationExportMode = 'light' | 'full'

export interface AnnotationPayload {
  label: AnnotationLabel
  anomaly_type?: string | null
  note?: string | null
  source_log_file_id?: number | null
}

export interface Annotation {
  id: number
  slice_window_id: number
  source_log_file_id?: number | null
  binding_label?: string | null
  label: AnnotationLabel
  anomaly_type: string | null
  note: string | null
  created_at: string
  updated_at: string
}

export type AnnotationSortField = 'id' | 'label' | 'anomaly_type' | 'created_at' | 'updated_at' | 'window_start_ts'
export type AnnotationSortOrder = 'asc' | 'desc'

export interface AnnotationListItem {
  id: number
  package_id: number
  task_id: number
  window_id: number
  window_start_ts: number
  window_end_ts: number
  label: AnnotationLabel
  anomaly_type: string | null
  note: string | null
  created_at: string
  updated_at: string
}

export interface AnnotationListResponse {
  items: AnnotationListItem[]
  total: number
  page: number
  page_size: number
}

export interface AnnotationStatsResponse {
  total_annotations: number
  normal_count: number
  abnormal_count: number
}

export interface PendingAnnotationItem {
  package_id: number
  package_name?: string | null
  task_id: number
  task_name?: string | null
  window_id: number
  window_start_ts: number
  window_end_ts: number
  line_count: number
  file_count: number
  cpu_count: number
  module_count: number
}

export type PendingSortField = 'window_start_ts' | 'line_count'
export type PendingSortOrder = 'asc' | 'desc'

export interface PendingAnnotationFilters {
  package_id?: number
  task_id?: number
  start_ts?: number
  end_ts?: number
  min_line_count?: number
  max_line_count?: number
  keyword?: string
  sort_by?: PendingSortField
  sort_order?: PendingSortOrder
}

export interface PendingAnnotationListResponse {
  items: PendingAnnotationItem[]
  total: number
  page: number
  page_size: number
}

export interface AnnotationWorkbenchItem {
  annotation_id: number
  package_id: number
  task_id: number
  window_id: number
  window_start_ts: number
  window_end_ts: number
  label: AnnotationLabel
  anomaly_type: string | null
  note: string | null
  updated_at: string
}

export interface AnnotationWorkbenchResponse {
  items: AnnotationWorkbenchItem[]
  total: number
  page: number
  page_size: number
}

export interface AnnotationExportResponse {
  format: 'csv' | 'json'
  scope: AnnotationExportScope
  mode: AnnotationExportMode
  file_path: string
  file_name: string
  download_url: string
  item_count: number
}

File: jj/code_9.17/annotation-app/backend/app/schemas/annotation.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field, field_validator

LabelType = Literal["normal", "abnormal"]
ExportScope = Literal["current_filter", "all"]
ExportMode = Literal["light", "full"]


class AnnotationCreateRequest(BaseModel):
    label: LabelType
    anomaly_type: str | None = Field(default=None, max_length=128)
    note: str | None = Field(default=None, max_length=20000)
    # None => whole-window annotation; a source file id => file-level annotation.
    source_log_file_id: int | None = None

    @field_validator("anomaly_type")
    @classmethod
    def _normalize_anomaly_type(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None

    @field_validator("note")
    @classmethod
    def _normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class AnnotationUpdateRequest(BaseModel):
    label: LabelType
    anomaly_type: str | None = Field(default=None, max_length=128)
    note: str | None = Field(default=None, max_length=20000)

    @field_validator("anomaly_type")
    @classmethod
    def _normalize_anomaly_type(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None

    @field_validator("note")
    @classmethod
    def _normalize_note(cls, value: str | None) -> str | None:
        if value is None:
            return None
        stripped = value.strip()
        return stripped or None


class AnnotationResponse(BaseModel):
    id: int
    slice_window_id: int
    source_log_file_id: int | None = None
    # Human-readable binding target: the file's logical_path, or None for whole-window.
    binding_label: str | None = None
    label: LabelType
    anomaly_type: str | None
    note: str | None
    created_at: datetime
    updated_at: datetime


class AnnotationListItem(BaseModel):
    id: int
    package_id: int
    task_id: int
    window_id: int
    window_start_ts: float
    window_end_ts: float
    label: LabelType
    anomaly_type: str | None
    note: str | None
    created_at: datetime
    updated_at: datetime


class AnnotationListResponse(BaseModel):
    items: list[AnnotationListItem]
    total: int
    page: int
    page_size: int


class AnnotationStatsResponse(BaseModel):
    total_annotations: int
    normal_count: int
    abnormal_count: int


class PendingAnnotationItem(BaseModel):
    package_id: int
    package_name: str | None = None
    task_id: int
    task_name: str | None = None
    window_id: int
    window_start_ts: float
    window_end_ts: float
    line_count: int = 0
    file_count: int = 0
    cpu_count: int = 0
    module_count: int = 0


class PendingAnnotationListResponse(BaseModel):
    items: list[PendingAnnotationItem]
    total: int
    page: int
    page_size: int


class AnnotationWorkbenchItem(BaseModel):
    annotation_id: int
    package_id: int
    task_id: int
    window_id: int
    window_start_ts: float
    window_end_ts: float
    label: LabelType
    anomaly_type: str | None
    note: str | None
    updated_at: datetime


class AnnotationWorkbenchResponse(BaseModel):
    items: list[AnnotationWorkbenchItem]
    total: int
    page: int
    page_size: int


class AnnotationExportResponse(BaseModel):
    format: Literal["csv", "json"]
    scope: ExportScope
    mode: ExportMode
    file_path: str
    file_name: str
    download_url: str
    item_count: int


class AnnotationExportLightItem(BaseModel):
    package_id: int
    package_name: str
    task_id: int
    task_name: str
    window_id: int
    window_start_at: str
    window_end_at: str
    line_count: int
    file_count: int
    cpu_count: int
    module_count: int
    annotation: dict[str, Any]
    window_context: dict[str, Any]
